src/content/blog/hybrid-rendering.mdx

```diff
@@ -0,0 +1,114 @@
+import hybridBuild from '/src/assets/blog/hybrid-rendering/hybrid.png'
+
+Since the initial launch of Astro 1.0, developers have been able to choose between generating [static (SSG)](https://docs.astro.build/en/core-concepts/routing/#static-ssg-mode) or [server (SSR)](https://docs.astro.build/en/guides/server-side-rendering/) output. Static websites offer incredible performance by building your pages ahead-of-time. Servers leverage their dynamic power to generate HTML on-demand, unique for every request.
+
+Until now, each Astro project has required an all-or-nothing decision—is this fully static or should I deploy a server?
+
+[Astro 2.0](/astro-2) unlocks the best of both worlds with **hybrid rendering**. Choose between the performance of static pages or the full flexibility of server rendering for each page of your website.
+
+In this post, we'll go over how hybrid rendering works by exploring just a few of the new features and performance improvements made possible in Astro 2.0.
+
+## How does `astro build` work?
+
+We'll simplify things a bit for clarity, but we first need to understand that Astro's build process happens in multiple stages. First, Astro uses [Vite](https://vitejs.dev) to generate a server-side JavaScript bundle. While this bundle is being generated, we perform static analysis on your source code to generate a **client manifest**. The output of this bundle includes:
+
+-   **Server-side JavaScript** used to render HTML
+-   A **client manifest** which marks any components needed for client-side interactivity
+-   **CSS and other assets** for the client
+
+Next, Astro uses the **client manifest** to initiate a second build process which bundles optimized client-side JavaScript.
+
+If your configured [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) is `static`, Astro will execute the server-side JavaScript and write the output to `.html` files. The server-side JavaScript is then discarded.
+
+<BlogContentImage
+    src={staticBuild}
+    alt="Flow diagram of Astro's static build process"
+/>
+
+If your configured [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) is `server`, Astro passes the server-side JavaScript to an [`adapter`](https://docs.astro.build/en/reference/configuration-reference/#adapter) for further processing.
+Adapters ensure that your server-side JavaScript is compatible with a particular hosting provider's JavaScript runtime. Note that in this case, the final output is not a set of `.html` files, but the JavaScript code necessary to render HTML.
+
+<BlogContentImage
+    src={serverBuild}
+    alt="Flow diagram of Astro's server build process"
+/>
+
+## How does hybrid rendering work?
+
+In Astro 2.0, we've revamped the `astro build` pipeline to handle hybrid rendering.
+During the initial bundling process, we've added an additional static analysis step that determines which pages should be prerendered.
```

Comment:
Suggestion to avoid “we’ve” repeated across sentences and slightly punchier sentence:

```suggestion
During the initial bundling process, a new static analysis step determines which pages should be prerendered.
```